docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+  ) 1 2
+usage2 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 3 4
+```
+
+**To summarize, inlining for the sake of inlining is a great way to unnecessarily increase your program's bundle size**.
+
+#### Rewrite Rules
+
+A **rewrite rule** will replace one expression, usually involving literal values, with an equivalent expression. For example, an expression like `1 + 2` will always produce the value `3`. So, we can replace that expression with `3`. The rewrite rule here could be defined as, "Whenever the `+` function is called on two literal `Int` values, replace that expression with the sum of the two `Int` values."
+
+```purs
+-- Before
+foo = 1 + 2
+  --  ^^^^^ Rewrite rule: "I can optimize that! `1 + 2` is `3`"
+
+-- After
+foo = 3
+```
+
+However, some expressions will prevent such rewrite rules from triggering. For example:
+
+```purs
+foo =
+  let
+    a = { bar: 42 }
+  in
+    1 + a.bar
+--  ^^^^^^^^^ Rewrite rule: "I can't optimize that! `a.bar` is not a literal `Int` value."
+```
+
+Because the rewrite rule cannot see the literal `Int` value represented by `a.bar`, its optimization does not trigger. To solve this problem, we need to make it visible. The solution is to duplicate code via inlining.
+
+**To summarize, inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise hidden from its eyes.**
+
+#### Code Optimzation Example via Rewrite Rules and Inlining
+
+Let's see how this works in practice using this example:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = ignoreArgs 8 9
+```
+
+Let's assume `ignoreArgs` has an inline directive of `arity=1`. Here's what happens when we optimize this code piece.
+
+First, we see that `ignoreArgs` has 2 arguments applied. Since `2` >= `1`, the arity of the directive we specified, we inline `ignoreArgs`. Since `8` corresponds to `arg1`, we replace `arg1` with `8` when inlining `ignoreArgs`. We don't remove the original `ignoreArgs` declaration.
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = (\arg2 -> do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 8 + arg2
+  ) 9
+```
+
+At this point, `purs-backend-es`' default inliners will see a lambda (i.e. a "producer") being immediately applied to an argument (i.e. a "consumer"). Thus, it will inline that argument into the lambda's body. This gets us:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 8 + 9
+```
+
+A rewrite rule will detect the `8 + 9` expression, and replace it with its sum: `17`. However, `1 + a.bar` and `a.bar + 17` don't trigger the rewrite rule because one of the values is not a literal `Int` value.
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 17
```

Comment:
This example is somewhat confusing because `+` is commutative. That is, if the optimizer was smart enough it could see that it's equivalent to `a.bar + 18`. In reality, it only currently tries to evaluate around abstract terms associatively, and even then only to a fixed depth, because doing this in general is quadratic.

So the example is accurate, but it invites more questions from the reader that I don't think you want to answer here.